content/docs/user-guide/data-pipelines/defining-pipelines.md

```diff
@@ -0,0 +1,204 @@
+      - src/preprocess.py
+    outs:
+      - data/preprocessed
+```
+
+<admon type="info">
+
+We use [GNU/Linux](https://www.gnu.org/software/software.html) in our examples,
+but Windows or other shells can be used too. Note that while DVC is
+platform-agnostic, the commands you define may have environment-specific
+requirements.
+
+</admon>
+
+You can write the `dvc.yaml` file above directly, or DVC can do it with the
+following call:
+
+```dvc
+$ dvc stage add --name preprocess \
+                --deps src/preprocess.py \
+                --deps data/raw \
+                --outs data/preprocessed \
+                python src/preprocess.py
+```
+
+<admon type="tip">
+
+One advantage of using `dvc stage add` is that it will verify the validity of
+the arguments provided (otherwise they won't be checked until execution). A
+disadvantage is that advanced pipelining features such as [templating] are not
+always available this way.
+
+</admon>
+
+[tracked with dvc]: /doc/start/data-management
+[templating]: /doc/user-guide/project-structure/pipelines-files#templating
+
+DVC writes lock files (`dvc.lock`) to complement `dvc.yaml` operations.
+Versioning both files with Git brings your pipelining process onto standard Git
+workflows.
+
+<details>
+
+### `dvc.lock` files: click to learn more.
+
+<admon type="info">
+
+You should never need to see the contents `dvc.lock` for regular DVC work.
+
+</admon>
+
+Lock files help DVC fix the state of the pipeline as it was last executed in
+order to compare it against the current state of the <abbr>workspace</abbr>. The
+following sample shows the kind of details saved.
+
+```yaml
+schema: '2.0'
+stages:
+  preprocess:
+    cmd: src/preprocess.py
+    deps:
+      - path: data/raw
+        md5: 687552951726b99c2eee15d29b4ccf0e
+        size: 17397976
+      - path: src/preprocess.py
+        md5: 51627ab6d865c51a634959dbc4914d24
+        size: 14623
+    outs:
+      - path: data/preprocessed
+        md5: 21188b73b5661d4730d769f795462485.dir
+        size: 154683
+        nfiles: 312
+```
+
+</details>
+
+## Simple dependencies
+
+There's more than one type of stage dependency: A simple dependency is a file or
+directory needed for the stage `cmd` to run successfully. When it's contents
+have changed, DVC "invalidates" the stage -- it knows that it needs to run
+again.
+
+<admon type="info">
+
+DVC [calculates a hash] of file/dir contents to compare vs. previous versions.
+This is a distinctive mechanism over traditional build tools like `make`.
+
+[calculates a hash]:
+  /doc/user-guide/project-structure/internal-files#structure-of-the-cache-directory
+
+</admon>
+
+File system-level dependencies are defined in the `deps` list of `dvc.yaml`;
+Alternatively, using the `--deps` (`-d`) option of `dvc stage add` (see the
+previous section's example).
+
+<details>
+
+### External dependencies: click to learn more.
+
+A less common kind of dependency is a _URL dependency_. Instead of files in a
```

Comment:
better to describe `dvc import` here?

Reply:
Not sure. I'm mentioning both now, PTAL.